modules/utxo-lib/src/bitgo/wallet/Psbt.ts

```diff
@@ -489,6 +497,231 @@ export function addXpubsToPsbt(psbt: UtxoPsbt, rootWalletKeys: RootWalletKeys):
+ * @param psbt - The PSBT to extract root nodes from.
+ * @returns An array of the three root BIP32 keys, or undefined if none are present.
+ */
+export function getMultiSigRootNodes(psbt: UtxoPsbt): Triple<BIP32Interface> | undefined {
+  const bip32s = psbt.data.globalMap.globalXpub?.map((xpub) =>
+    BIP32Factory(eccLib).fromBase58(bs58check.encode(xpub.extendedPubkey))
+  );
+  assert(!bip32s || isTriple(bip32s), `Invalid globalXpubs in PSBT. Expected 3 or none. Got ${bip32s?.length}`);
+  return bip32s;
+}
+
+/**
+ * Maps a parsed 2-of-3 script type to its corresponding ScriptType2Of3 array.
+ * Used to handle the various script types a PSBT input or output can be part of.
+ * @param parsedScriptType - The parsed script type from a PSBT input/output.
+ * @returns An array of ScriptType2Of3 based on the parsedScriptType.
+ */
+export function toScriptType2Of3s(parsedScriptType: ParsedScriptType2Of3): ScriptType2Of3[] {
+  return parsedScriptType === 'taprootScriptPathSpend'
+    ? ['p2trMusig2', 'p2tr']
+    : parsedScriptType === 'taprootKeyPathSpend'
+    ? ['p2trMusig2']
+    : [parsedScriptType];
+}
+
+/**
+ * Checks if a specific permutation of public keys matches the provided scriptPubKey.
+ * Utilizes script type conversion and script creation to verify the match.
+ * @param params - An object containing the required parameters for script matching.
+ * @returns True if the scriptPubKey matches the constructed script for the given permutation and type.
+ */
+function matchesScript({
+  publicKeys,
+  perm,
+  scriptPubKey,
+  parsedScriptType,
+  network,
+}: {
+  publicKeys: Buffer[];
+  perm: Triple<number>;
+  scriptPubKey: Buffer;
+  parsedScriptType: ParsedScriptType2Of3;
+  network: Network;
+}): boolean {
+  const orderedPublicKeys: Triple<Buffer> = [publicKeys[perm[0]], publicKeys[perm[1]], publicKeys[perm[2]]];
+  const scriptTypes = toScriptType2Of3s(parsedScriptType);
+  return scriptTypes.some((scriptType) =>
+    createOutputScript2of3(orderedPublicKeys, scriptType, network).scriptPubKey.equals(scriptPubKey)
+  );
+}
+
+/**
+ * Determines the correct order of public keys for a 2-of-3 multisig input based on the scriptPubKey.
+ * It iterates through all possible permutations to find a match.
+ * @param publicKeys - The public keys involved in the multisig setup.
+ * @param scriptPubKey - The scriptPubKey from the PSBT input to match against.
+ * @param parsedScriptType - The type of script the PSBT input is using.
+ * @param network - The Bitcoin network the PSBT is for.
+ * @returns An ordered array of indices representing the correct order of public keys.
+ */
+function determineOrder(
+  publicKeys: Triple<Buffer>,
+  scriptPubKey: Buffer,
+  parsedScriptType: ParsedScriptType2Of3,
+  network: Network
+): Triple<number> {
+  const permutations: Array<Triple<number>> = [
+    [0, 1, 2],
+    [0, 2, 1],
+    [1, 0, 2],
+    [1, 2, 0],
+    [2, 0, 1],
+    [2, 1, 0],
+  ];
```

Comment:
instead of doing it by index we can do it by buffer

```
const [a, b, c] = publicKeys;
const permutations = [
 [a, b, c],
 [a, c, b],
 ...
]`
```

one less layer of indirection